### PR TITLE
[Inventory] Documentation of Inventory component

### PR DESCRIPTION
**Pull request by Arminek**
| Q                    | A
| -------------------| ---
| Doc fix?          | [yes]
| New docs?     | [no] 
| Fixed tickets   | [no]

### CODE REVIEW

File: components/Inventory/basic_usage.rst [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+.. code:: php
+
+    <?php
+
+    use Sylius\Component\Inventory\Operator\InventoryOperator;
+    use Sylius\Component\Inventory\Checker\AvailabilityChecker;
+    use Sylius\Component\Inventory\Operator\BackordersHandler;
+
+    $inventoryUnitRepository; //Repository model.
```
Comment (TheMadeleine) [inline components/Inventory/basic_usage.rst]: spaces after //

File: components/Inventory/basic_usage.rst [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+    $stockable; //Stockable model.
+    $eventDispatcher; //It gives possibilty to hook before or after each operation.
+    // If you are not familiar with events. Check symfony event dispatcher.
```
Comment (TheMadeleine) [inline components/Inventory/basic_usage.rst]: `If you are not familiar with events, check the Symfony Event Dispatcher.`

File: components/Inventory/basic_usage.rst [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    $inventoryOperator->hold($stockable, 4);
+    $stockable->getOnHold(); //Output will be 4.
+
+    $inventoryOperator->release($stockable, 3);
+    $stockable->getOnHold(); //Output will be 1.
+
+Decrease
+~~~~~~~~
+
+This specific case will be more complicated.
```
Comment (TheMadeleine) [inline components/Inventory/basic_usage.rst]: I'm not sure if such a description is enough here :)

File: components/Inventory/basic_usage.rst [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+    <?php
+
+    use Sylius\Component\Inventory\Operator\InventoryOperator;
+    use Sylius\Component\Inventory\Checker\AvailabilityChecker;
+    use Sylius\Component\Inventory\Operator\BackordersHandler;
+
+    $inventoryUnitRepository; //Repository model.
+    $stockable; //Stockable model.
+    $eventDispatcher; //It gives possibilty to hook before or after each operation.
```
Comment (TheMadeleine) [inline components/Inventory/basic_usage.rst]: It gives **a** possibility [...]

File: components/Inventory/basic_usage.rst [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+    $inventoryUnits->add($inventoryUnit1);
+    $inventoryUnits->add($inventoryUnit2);
+
+    count($inventoryUnits); //Output will be 2.
+    $inventoryOperator->decrease($inventoryUnits);
+    $stockable->getOnHand(); //Output will be 4.
+
+.. caution::
+    All methods in **InventoryOperator** throws `InvalidArgumentException`_ or `InsufficientStockException`_ if an error occurs.
```
Comment (TheMadeleine) [inline components/Inventory/basic_usage.rst]: *throw

File: components/Inventory/basic_usage.rst [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+
+.. _Symfony EventDispatcher: http://symfony.com/doc/current/components/event_dispatcher/introduction.html
+
+NoopInventoryOperator
+---------------------
+
+In some cases, you may want to have unlimited inventory, this operator will allow you to do that.
+
+.. hint::
+    This operator based on null object pattern. For more detailed information go to `Null Object pattern`_.
```
Comment (TheMadeleine) [inline components/Inventory/basic_usage.rst]: `This operator is based on the null object patter.`

File: components/Inventory/basic_usage.rst [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+
+BackordersHandler
+-----------------
+
+The **BackorderHandler** changes inventory unit state.
+
+Process backorders
+~~~~~~~~~~~~~~~~~~
+
+This method will change inventory unit state to ``backordered`` if quantity of requested inventory units will be insufficient.
```
Comment (TheMadeleine) [inline components/Inventory/basic_usage.rst]: [...] change **the** inventory unit state [...] if **the** quantity [...]

File: components/Inventory/basic_usage.rst [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+
+    use Sylius\Component\Inventory\Operator\BackordersHandler;
+
+    $inventoryUnitRepository; //Repository model.
+    $stockable; //Stockable model.
+
+    $backordersHandler = new BackordersHandler($inventoryUnitRepository);
+    $stockable->getOnHand(); //Output will be 6.
+
+    //Let assume that we got 9 inventory units with state 'backordered'.
```
Comment (TheMadeleine) [inline components/Inventory/basic_usage.rst]: `// Let's assume that we have 9 [...] `

File: components/Inventory/basic_usage.rst [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+.. note::
+    For more detailed information go to `Sylius API BackordersHandler`_.
+
+.. _Sylius API InventoryUnitFactory: http://api.sylius.org/Sylius/Component/Inventory/Operator/BackordersHandler.html
+
+
+AvailabilityChecker
+-------------------
+
+The **AvailabilityChecker** checks availability for given stockable object.
```
Comment (TheMadeleine) [inline components/Inventory/basic_usage.rst]: availability **of** a given stockable object

File: components/Inventory/basic_usage.rst [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+To charactrize an object which is an **AvailabilityChecker**, it needs to implement the :ref:`component_inventory_checker_availability_checker_interface`_.
+Second parameter in ``->isStockSufficient()`` method indicates that customer want a merchandise in specific quantity.
```
Comment (TheMadeleine) [inline components/Inventory/basic_usage.rst]: That sentence is a bit too sophisticated imo :)

`Second parameter of the isStockSufficient() method gives a possibility to check for a given quantity of a stockable.`

File: components/Inventory/basic_usage.rst [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+    $stockable->getOnHold(); //Output will be 4
+
+    $availabilityChecker = new AvailabilityChecker(false); //backorders = false;
+    $availabilityChecker->isStockAvailable($stockable); //Output will be true.
+    $availabilityChecker->isStockSufficient($stockable, 5); //Output will be false.
+
+Backorders
+~~~~~~~~~~
+
+The backorder property generally indicates that customer demand for a product or service
```
Comment (TheMadeleine) [inline components/Inventory/basic_usage.rst]: `[...] that the customer's demand`

File: components/Inventory/basic_usage.rst [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+
+    $stockable; //Stockable model.
+    $stockable->isAvailableOnDemand(); //Output will be true.
+
+    $availabilityChecker = new AvailabilityChecker(false); //backorders = false;
+    $availabilityChecker->isStockAvailable($stockable); //Output will be true.
+    $availabilityChecker->isStockSufficient($stockable, 5); //Output will be true.
+
+.. hint::
+    In above cases value of ``->getOnHand()`` and ``->getOnHold()`` will be irrelevant.
```
Comment (TheMadeleine) [inline components/Inventory/basic_usage.rst]: `In the above cases results of [...]`

File: components/Inventory/basic_usage.rst [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+
+.. note::
+    For more detailed information go to `Sylius API AvailabilityChecker`_.
+
+.. _Sylius API AvailabilityChecker: http://api.sylius.org/Sylius/Component/Inventory/Checker/AvailabilityChecker.html
+
+InventoryUnitFactory
+--------------------
+
+The **InventoryUnitFactory** basically is simple factory. It creates collection of new inventory units.
```
Comment (TheMadeleine) [inline components/Inventory/basic_usage.rst]: Thank you Mr. Obvious! :)

`The **InventoryUnitFactory** creates a collection of new inventory units.`

File: components/Inventory/interfaces.rst [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+Service Interfaces
+------------------
+
+.. _component_inventory_checker_availability-checker-interface:
+
+AvailabilityCheckerInterface
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This interface provides methods for checking availability of stockable objects.
+To characterize an object which is a **AvailabilityChecker** it needs to implement the ``AvailabilityCheckerInterface``
```
Comment (TheMadeleine) [inline components/Inventory/interfaces.rst]: I suggest this sentence is highly unnecessary!

File: components/Inventory/interfaces.rst [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+    For more detailed information go to `Sylius API AvailabilityCheckerInterface`_.
+
+.. _Sylius API AvailabilityCheckerInterface: http://api.sylius.org/Sylius/Component/Inventory/Checker/AvailabilityCheckerInterface.html
+
+.. _component_inventory_factory_inventory-unit-factory-interface:
+
+InventoryUnitFactoryInterface
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This interface is implemented by the service responsible for creating collection of new inventory units.
```
Comment (TheMadeleine) [inline components/Inventory/interfaces.rst]: try plural -> ` implemented by services responsible for`
and apply it for the ones below also.

File: components/Inventory/models.rst [hunk 16]
```diff
@@ -0,0 +1,30 @@
+Models
+======
+
+.. _component_inventory_model_inventory-unit:
+
+InventoryUnit
+-------------
+
+**InventoryUnit** object represents an inventory unit.
```
Comment (TheMadeleine) [inline components/Inventory/models.rst]: damn, wouldn't know that without that sentence. ;< :D

File: components/Inventory/state_machine.rst [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+.. tip::
+    Please keep in mind that these states are just default, you can define and use your own.
+    If you use this component with :doc:`/bundles/SyliusInventoryBundle/index` and Symfony2, you will have full state machine configuration at your disposal.
+
+.. _component_inventory_inventory-unit-transitions:
+
+Inventory Unit Transitions
+--------------------------
+
+There are following order's transitions by default:
```
Comment (TheMadeleine) [inline components/Inventory/state_machine.rst]: **the** following

File: components/Inventory/models.rst (hunk 16 above)
Comment (pamil) [inline components/Inventory/models.rst]: :tada: :D

File: components/Inventory/basic_usage.rst [hunk 18]
```diff
@@ -0,0 +1,378 @@
...
+
+    use Sylius\Component\Inventory\Factory\InventoryUnitFactory;
+    use Sylius\Component\Inventory\Model\InventoryUnitInterface;
+
+    $inventoryUnitRepository; // Repository model.
+    $product = new Product(); // Stockable model.
+
+    $inventoryUnitFactory = new InventoryUnitFactory($inventoryUnitRepository);
+
+    $inventoryUnits = $inventoryUnitFactory->create($product, 10, InventoryUnitInterface::STATE_RETURNED); // Output will be collection of inventory units.
```
Comment (tuka217) [inline components/Inventory/basic_usage.rst]: It is really long line. Maybe put comment above code.

File: components/Inventory/interfaces.rst (hunk 14 above)
Comment (pjedrzejewski) [inline components/Inventory/interfaces.rst]: :+1:

File: components/Inventory/models.rst (hunk 16 above)
Comment (pjedrzejewski) [inline components/Inventory/models.rst]: :octopus:

**Comment by pjedrzejewski** [comment]
:+1: I think this is enough doc for now. The component itself needs to be reworked, so we will have much nicer docs after few changes to the code itself. Good work. :+1:

**Comment by pjedrzejewski** [comment]
:+1:

**Pull request merged**